bundles/org.openhab.automation.pythonscripting/src/main/java/org/openhab/automation/pythonscripting/internal/PythonScriptEngine.java

```diff
@@ -0,0 +1,508 @@
+                .option(PYTHON_OPTION_POSIXMODULEBACKEND, "java") //
+                // Force to automatically import site.py module, to make Python packages available
+                .option(PYTHON_OPTION_FORCEIMPORTSITE, Boolean.toString(true)) //
+                // The sys.executable path, a virtual path that is used by the interpreter
+                // to discover packages
+                .option(PYTHON_OPTION_EXECUTABLE,
+                        PythonScriptEngineFactory.PYTHON_DEFAULT_PATH.resolve("bin").resolve("python").toString())
+                // Set the python home to be read from the embedded resources
+                .option(PYTHON_OPTION_PYTHONHOME, PythonScriptEngineFactory.PYTHON_DEFAULT_PATH.toString()) //
+                // Set python path to point to sources stored in
+                .option(PYTHON_OPTION_PYTHONPATH,
+                        PythonScriptEngineFactory.PYTHON_LIB_PATH.toString() + File.pathSeparator
+                                + PythonScriptEngineFactory.PYTHON_DEFAULT_PATH.toString())
+                // pass the path to be executed
+                .option(PYTHON_OPTION_INPUTFILEPATH, PythonScriptEngineFactory.PYTHON_DEFAULT_PATH.toString()) //
+                // make sure the TopLevelExceptionHandler calls the excepthook to print Python exceptions
+                .option(PYTHON_OPTION_ALWAYSRUNEXCEPTHOOK, Boolean.toString(true)) //
+                // emulate jython behavior (will slowdown the engine)
+                .option(PYTHON_OPTION_EMULATEJYTHON,
+                        String.valueOf(pythonScriptEngineConfiguration.isJythonEmulation()));
+
+        if (pythonScriptEngineConfiguration.isCachingEnabled()) {
+            contextConfig.option(PYTHON_OPTION_DONTWRITEBYTECODEFLAG, Boolean.toString(false)) //
+                    .option(PYTHON_OPTION_CACHEDIR, PYTHON_CACHEDIR_PATH);
+        } else {
+            contextConfig.option(PYTHON_OPTION_DONTWRITEBYTECODEFLAG, Boolean.toString(true)) //
+                    // causes the interpreter to always assume hash-based pycs are valid
+                    .option(PYTHON_OPTION_CHECKHASHPYCSMODE, "never");
+        }
+
+        delegate = GraalPythonScriptEngine.create(ENGINE, contextConfig);
+    }
+
+    @Override
+    protected void beforeInvocation() {
+        lock.lock();
+        logger.debug("Lock acquired before invocation.");
+
+        if (initialized) {
+            return;
+        }
+
+        // needed to get ThreadGroups working in JDK17 (openhab4). Otherwise, Python threads are not working properly.
+        // - if ThreadGroups.isDaemon == true and last python thread is done
+        // ==> ThreadGroup is closed and new python threads can't started anymore
+        // - in jdk21 is handled differently
+        // ==> ThreadGroups does not use the daemon flag anymore and ThreadGroups are not closable anymore
+        // - can be removed in jdk21 only environments
```

Comment:
Can we remove this here hence?
main branch is openHAB 5 which requires JDK 21. I know you want to provide a 4.x version of the add-on, but this ideally shouldn't pollute the code on main.

Reply:
currently, if I build the openhab4 version, I just copy over the source code and compile. If this is a blocker for you, I can remove it.

My current plant is to keep it until openhab 5 is released. Until this time I will also compile specific openhab4 versions. So I would remove it in a openhab 5.1 branch.

Reply:
I would prefer to remove it here, and then add it when compiling the 4.x version. You can keep a copy of the 4.3.x branch with the Python add-on and an extra commit containing this change, and whenever the Python binding changes on main, rebase it.